src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -106,215 +125,272 @@ class column_pruner {
-      //  3. If the repeated field is a group with one field and is named either array or uses the
-      //     LIST-annotated group's name with _tuple appended then the repeated type is the element
-      //     type and elements are required.
-      //  4. Otherwise, the repeated field's type is the element type with the repeated field's repetition.
-
-      // MAP is used to annotate types that should be interpreted as a map from keys to values. MAP must
-      // annotate a 3-level structure:
-      //
-      //  * The outer-most level must be a group annotated with MAP that contains a single field named
-      //    key_value. The repetition of this level must be either optional or required and determines
-      //    whether the list is nullable.
-      //  * The middle level, named key_value, must be a repeated group with a key field for map keys
-      //    and, optionally, a value field for map values.
-      //  * The key field encodes the map's key type. This field must have repetition required and must
-      //    always be present.
-      //  * The value field encodes the map's value type and repetition. This field can be required,
-      //    optional, or omitted.
-      //
-      // It is required that the repeated group of key-value pairs is named key_value and that its
-      // fields are named key and value. However, these names may not be used in existing data and
-      // should not be enforced as errors when reading.
-      //
-      // Some existing data incorrectly used MAP_KEY_VALUE in place of MAP. For backward-compatibility,
-      // a group annotated with MAP_KEY_VALUE that is not contained by a MAP-annotated group should be
-      // handled as a MAP-annotated group.
-
-      // Parquet says that the map's value is optional, but Spark looks like it would throw an exception
-      // if it ever actually saw that in practice, so we should too.
-      CUDF_FUNC_RANGE();
-      // The maps are sorted so we can compress the tree...
-      // These are the outputs of the computation
-      std::map<int, int, std::less<int>> chunk_map;
-      std::map<int, int, std::less<int>> schema_map;
-      std::map<int, int, std::less<int>> num_children_map;
-      // Start off with 0 children in the root, will add more as we go
-      schema_map[0] = 0;
-      num_children_map[0] = 0;
-
-      // num_children_stack and tree_stack hold the current state as we walk though schema
-      std::vector<int> num_children_stack;
-      std::vector<column_pruner*> tree_stack;
-      tree_stack.push_back(this);
-      if (schema.size() == 0) {
-        throw std::invalid_argument("a root schema element must exist");
+    void filter_schema(std::vector<parquet::format::SchemaElement> & schema, const bool ignore_case,
+            std::size_t & current_input_schema_index, std::size_t & next_input_chunk_index,
+            std::vector<int> & chunk_map, std::vector<int> & schema_map, std::vector<int> & schema_num_children) {
```

Comment:
A "good" practice is to split such giant function into multiple smaller functions (like `filter_schema_value`, `filter_schema_struct`, etc.). Not necessarily required here but it's recommended to do so.